src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -7982,246 +7982,263 @@ void StorageReplicatedMergeTree::replacePartitionFrom(
-
-    LOG_DEBUG(log, "Cloning {} parts", src_all_parts.size());
-
     static const String TMP_PREFIX = "tmp_replace_from_";
     auto zookeeper = getZooKeeper();
 
-    /// Retry if alter_partition_version changes
-    for (size_t retry = 0; retry < 1000; ++retry)
+    std::unordered_set<String> partitions;
+    if (partition->as<ASTPartition>()->all)
     {
-        DataPartsVector src_parts;
-        MutableDataPartsVector dst_parts;
-        std::vector<scope_guard> dst_parts_locks;
-        Strings block_id_paths;
-        Strings part_checksums;
-        std::vector<EphemeralLockInZooKeeper> ephemeral_locks;
-        String alter_partition_version_path = zookeeper_path + "/alter_partition_version";
-        Coordination::Stat alter_partition_version_stat;
-        zookeeper->get(alter_partition_version_path, &alter_partition_version_stat);
-
-        /// Firstly, generate last block number and compute drop_range
-        /// NOTE: Even if we make ATTACH PARTITION instead of REPLACE PARTITION drop_range will not be empty, it will contain a block.
-        /// So, such case has special meaning, if drop_range contains only one block it means that nothing to drop.
-        /// TODO why not to add normal DROP_RANGE entry to replication queue if `replace` is true?
-        MergeTreePartInfo drop_range;
-        std::optional<EphemeralLockInZooKeeper> delimiting_block_lock;
-        bool partition_was_empty = !getFakePartCoveringAllPartsInPartition(partition_id, drop_range, delimiting_block_lock, true);
-        if (replace && partition_was_empty)
-        {
-            /// Nothing to drop, will just attach new parts
-            LOG_INFO(log, "Partition {} was empty, REPLACE PARTITION will work as ATTACH PARTITION FROM", drop_range.partition_id);
-            replace = false;
-        }
-
-        if (!replace)
-        {
-            /// It's ATTACH PARTITION FROM, not REPLACE PARTITION. We have to reset drop range
-            drop_range = makeDummyDropRangeForMovePartitionOrAttachPartitionFrom(partition_id);
-        }
-
-        assert(replace == !LogEntry::ReplaceRangeEntry::isMovePartitionOrAttachFrom(drop_range));
-
-        scope_guard intent_guard;
         if (replace)
-        {
-            queue.addDropReplaceIntent(drop_range);
-            intent_guard = scope_guard{[this, my_drop_range = drop_range]() { queue.removeDropReplaceIntent(my_drop_range); }};
+            throw DB::Exception(ErrorCodes::SUPPORT_IS_DISABLED, "Only support DROP/DETACH/ATTACH PARTITION ALL currently");
+
+        partitions = src_data.getAllPartitionIds();
+        LOG_INFO(log, "Will try to attach {} partitions without replace", partitions.size());
+    } else
+    {
+        partitions = std::unordered_set<String>();
+        partitions.emplace(getPartitionIDFromQuery(partition, query_context));
+    }
+
+    for (const auto & partition_id : partitions)
```

Comment:
Since it will not be atomic anyway, let's create another function that iterates through partitions and calls `replacePartitionFrom` for each partition, and then calls `waitForLogEntryToBeProcessedIfNecessary` for each created entry

Reply:
please, refine what do you mean, and why

Reply:
Yes, but not only for readability. Currently, the loop works like this:
 1. process partition A (other replicas do nothing)
 2. wait for other replicas
 3. process partition B (other replicas do nothing)
 4. wait for other replicas
 5. process partition C (other replicas do nothing)
 6. wait for other replicas

And with the change that I suggested it will be:
 1. process partition A
 2. process partition B (other replicas can process partition A in the meantime)
 3. process partition C (other replicas can process partitions A and B in the meantime)
 4. wait for other replicas (waiting for A and B will probably be a no-op, so we will wait much less)

Reply:
Got it, thanks. Nice remark